onnx/defs/math/defs.cc

```diff
@@ -1226,20 +1236,144 @@ Computes the error function of the given input tensor element-wise.
+            "Constrain input b and its zero point data type to 8-bit integer tensor.")
+        .TypeConstraint(
+            "T3",
+            {"tensor(int8)", "tensor(uint8)"},
+            "Constrain output y and its zero point data type to 8-bit integer tensor.")
+        .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext&
+                                              ctx) {
+          auto a_type = ctx.getInputType(0);
+          auto b_type = ctx.getInputType(3);
+          auto y_type = ctx.getOutputType(0);
+          if (nullptr == a_type || nullptr == b_type || nullptr == y_type ||
+              a_type->value_case() != ONNX_NAMESPACE::TypeProto::kTensorType ||
+              b_type->value_case() != ONNX_NAMESPACE::TypeProto::kTensorType) {
+            fail_type_inference(
+                "inputs are expected to have tensor type and output type should not be null.");
+          }
+
+          if (ONNX_NAMESPACE::TensorProto::UINT8 ==
+                  a_type->tensor_type().elem_type() &&
+              ONNX_NAMESPACE::TensorProto::UINT8 ==
+                  b_type->tensor_type().elem_type()) {
+            y_type->mutable_tensor_type()->set_elem_type(
+                ONNX_NAMESPACE::TensorProto::UINT8);
+          } else {
+            y_type->mutable_tensor_type()->set_elem_type(
+                ONNX_NAMESPACE::TensorProto::INT8);
+          }
+
+          matmulShapeInference(ctx, 0, 3);
+        }));
+
+static const char* MatMulInteger_ver10_doc = R"DOC(
+Matrix product that behaves like numpy.matmul: https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.matmul.html.
+The production MUST never overflow. The accumulation may overflow if and only if in 32 bits.
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    MatMulInteger,
+    10,
+    OpSchema()
+        .SetDoc(MatMulInteger_ver10_doc)
+        .Input(0, "A", "N-dimensional matrix A", "T1")
+        .Input(1, "B", "N-dimensional matrix B", "T2")
+        .Input(
+            2,
+            "a_zero_point",
+            "Zero point tensor for input 'A'. It's optional and default value is 0. It could be a scalar or a 1-D tensor, "
+            "which means a per-tensor or per-row quantization. If it's a 1-D tensor, its number of elements "
+            "should be equal to the number of rows of input 'A'.",
+            "T1",
+            OpSchema::Optional)
+        .Input(
+            3,
+            "b_zero_point",
+            "Scale tensor for input 'B'. It's optional and default value is 0.  It could be a scalar or a 1-D tensor, "
+            "which means a per-tensor or per-column quantization. If it's a 1-D tensor, its number "
+            "of elements should be equal to the number of columns of input 'B'.",
+            "T2",
+            OpSchema::Optional)
+        .Output(0, "Y", "Matrix multiply results from A * B", "T3")
+        .TypeConstraint(
+            "T1",
+            {"tensor(int8)", "tensor(uint8)"},
```

Comment:
Please consider support int16 as well. There's cblas_gemm_s16s16s32 for int16 inputs to generate int32 outputs.

Reply:
@darrenscrews is going to make that (more types support) in separate PR.